### DISCUSSION TITLE
Simple shell script for making managing final kernel boot parameters a bit easier...

### DISCUSSION THREAD

**Discussion by Matthew-Tate-Scarbrough** [zbm-dev/zfsbootmenu#123] [General]
So anyway, after asking the other day on how to actually manage the kernel boot parameters, I knew I would be doing it a lot, so I figured I would write a script for me to be able to do that much easier.  Even if it is a one-liner, if you have to do it twice, you may thrice, so it is best to write a script and document it.

I only have a year on \*NIX, in general's, experience.  I tried to keep the script as POSIX as I know how.  It's probably not the best.  Here is the link to it on my repo in which I keep my scripts: [zbm_kernel_params](https://github.com/Matthew-Tate-Scarbrough/universalricing/blob/master/.local/bin/zbm_kernel_params).

In short, it reads `/proc/cmdline` to get the current kernel parameters.  If you simply run the shell script as a oneliner, it appends the new kernel parameters:

1. `zbm_kernel_params new_param1` \
— or —
2. `zbm_kernel_params "new_param1 new_param2"`

Alternatively, it has an interactive mode, ie.

    zbm_kernel_params
    Enter parameters> new_param1 new_param2

And it can create a temporary file, `/tmp/old_kern_params_tmp`, and which it then forces the end user to edit with `vi`.  Upon exit, these are applied and the temporary file is deleted.

    zbm_kernel_params -e

And lastly, it can display the current boot parameters that should take effect (or remain in effect) at next reboot:

    zbm_kernel_params -v

- - - -

I doubt it would be something you would be interested in cloning (and perhaps making a bit more straight-forward) for the project, but I am at least throwing it out there.  I have tested it how I could think to, I used `shellcheck`.  It should work as-is.  The only manual intervention required would be for the end user to define their ZFS root dataset.  The guide for Void on the wiki uses `zroot/ROOT`. I used `zroot/ROOT`.  I think FreeBSD uses `zroot/ROOT` (last time I played with FBSD 12.)

**Comment by zdykstra** [comment]
Thank you for taking the time to both write this helper and to write up the detailed discussion on it!  I like the ability to append parameters to the command line, as well as being able to edit it easily in vi/$EDITOR.

If you're open to a few suggestions on small changes to the script:

* Instead of using static temporary files, use `mktemp`to create a unique temporary file on each invocation.
* Add an exit TRAP that tests for the temporary files and deletes them, so a ctrl-c doesn't leave them around
* Instead of using `vi`, test for `$EDITOR` and use that, otherwise default to vi. 
* It might be nice to add a line return on whitespace when showing the `org.zfsbootmenu:commandline` value, to make it readable in bite-sized chunks, instead of one long line.

With some small changes (BASH-isms are fine, ZFSBootMenu has a hard dependency on Bash), this would fit nicely in the `contrib/` tools section.
